src/common/walFilter/versions/recordProcessGPDB7.c

```diff
@@ -0,0 +1,458 @@
+ * include origin.h in a bunch of xlog related places.
+ */
+typedef uint16 RepOriginId;
+
+typedef struct XLogRecordDataHeaderShort
+{
+    uint8_t id;                   /* XLR_BLOCK_ID_DATA_SHORT */
+    uint8_t data_length;          /* number of payload bytes */
+} __attribute__((packed)) XLogRecordDataHeaderShort;
+
+typedef struct XLogRecordDataHeaderLong
+{
+    uint8_t id;                   /* XLR_BLOCK_ID_DATA_LONG */
+    uint32_t data_length;         /* number of payload bytes */
+} __attribute__((packed)) XLogRecordDataHeaderLong;
+
+static void
+XLogRecordGPDB7ToLog(const XLogRecordGPDB7 *const this, StringStatic *const debugLog)
+{
+    strStcFmt(
+        debugLog,
+        "{xl_tot_len: %u, xl_xid: %u, xl_prev: %" PRIu64 ", xl_info: %u, xl_rmid: %u, xl_crc: %u}",
+        this->xl_tot_len,
+        this->xl_xid,
+        this->xl_prev,
+        this->xl_info,
+        this->xl_rmid,
+        this->xl_crc
+        );
+}
+
+#define FUNCTION_LOG_XLOG_RECORD_GPDB7_TYPE                                                                                        \
+    XLogRecordGPDB7 *
+#define FUNCTION_LOG_XLOG_RECORD_GPDB7_FORMAT(value, buffer, bufferSize)                                                           \
+    FUNCTION_LOG_OBJECT_FORMAT(value, XLogRecordGPDB7ToLog, buffer, bufferSize)
+
+FN_EXTERN void
+validXLogRecordHeaderGPDB7(const XLogRecordBase *recordBase, __attribute__((unused)) PgPageSize heapPageSize)
+{
+    const XLogRecordGPDB7 *const record = (const XLogRecordGPDB7 *const) recordBase;
+
+    if (record->xl_tot_len < sizeof(XLogRecordGPDB7))
+    {
+        THROW_FMT(FormatError, "invalid record length: wanted %zu, got %u", sizeof(XLogRecordGPDB7), record->xl_tot_len);
+    }
+    if (record->xl_rmid > RM_MAX_ID)
+    {
+        THROW_FMT(FormatError, "invalid resource manager ID");
+    }
+}
+
+FN_EXTERN void
+validXLogRecordGPDB7(const XLogRecordBase *const recordBase, __attribute__((unused)) const PgPageSize heapPageSize)
+{
+    const XLogRecordGPDB7 *const record = (const XLogRecordGPDB7 *const) recordBase;
+
+    if (record->xl_crc != xLogRecordChecksumGPDB7(record))
+    {
+        THROW_FMT(FormatError, "incorrect resource manager data checksum in record");
+    }
+}
+
+FN_EXTERN uint32_t
+xLogRecordHeaderSizeGPDB7(void)
+{
+    return sizeof(XLogRecordGPDB7);
+}
+
+FN_EXTERN uint32_t
+xLogRecordRmidSizeGPDB7(void)
+{
+    return offsetof(XLogRecordGPDB7, xl_rmid) + SIZE_OF_STRUCT_MEMBER(XLogRecordGPDB7, xl_rmid);
+}
+
+FN_EXTERN bool
+xLogRecordIsWalSwitchGPDB7(const XLogRecordBase *recordBase)
+{
+    const XLogRecordGPDB7 *const record = (const XLogRecordGPDB7 *const) recordBase;
+    return record->xl_rmid == RM_XLOG_ID && record->xl_info == XLOG_SWITCH;
+}
+
+static const RelFileNode *
+getRelFileNodeFromMainData(const XLogRecordGPDB7 *const record, const void *const mainData)
+{
+    uint8_t info = (uint8_t) (record->xl_info & ~XLR_INFO_MASK);
```

Comment:
record->xl_info is declared as uint8. uint8 is the same as uint8_t. Do we need casting to uint8_t?
```suggestion
    uint8 info = record->xl_info & ~XLR_INFO_MASK;
```

Reply:
After bit operations, the type becomes int, so we need to do the type cast again.